### ISSUE TITLE
OpenRCT2 may not work, if data resides on XFS/ReiserFS volume

### ISSUE THREAD

**Issue by janisozaur** [OpenRCT2/OpenRCT2#6268]
According to the documentation: https://www.gnu.org/software/libc/manual/html_node/Directory-Entries.html

> Only some filesystems have full support to return the type of the file, others might always return [`DT_UNKNOWN`].

We rely on unsupported behaviour in here:

https://github.com/OpenRCT2/OpenRCT2/blob/c6e162e31c9630edc221bc97cb72542b29176432/src/openrct2/platform/posix.c#L470-L475

XFS and ReiserFS may not carry this information.

We should instead check result of `S_ISDIR`, like we do in here: https://github.com/OpenRCT2/OpenRCT2/blob/c6e162e31c9630edc221bc97cb72542b29176432/src/openrct2/platform/posix.c#L133

**Comment by IntelOrca** [comment]
What about here?

https://github.com/OpenRCT2/OpenRCT2/blob/cf6f497cdeffc3d1a55e2eab6cd6218fe62421e4/src/openrct2/core/FileScanner.cpp#L329-L339

**Comment by janisozaur** [comment]
Yes, that too

**Comment by janisozaur** [comment]
https://linux-tips.com/t/readdir-function-on-xfs-filesystem-not-working-properly/115/2

As this is introducing syscall overheads, we may try to implement it only as fallback if `d_type == DT_UNKOWN`

**Comment by jchv** [comment]
Would it make sense to implement an abstraction around `scandir` and other primitives instead? Could make it a platform abstraction and implement it with `FindFirstFile` and `FindNextFile` on Windows instead. I guess it doesn't _have_ to be a platform abstraction, but it may be useful to do so anyways, for platforms that may not have full libc implementations, and plus, it feels a bit ugly to be passing around `dirent`s and handling `DT_UNKNOWN` at every possible invocation. Also, there may be unforeseen future benefits to adding an abstraction here, though I'm not sure.

Thoughts? Hope this makes sense.  I'm new here.

tl;dr: My plan would be to abstract filesystem calls as part of the platform layer and make other code not directly touch `dirent`s or deal with `DT_UNKNOWN`.

**Comment by IntelOrca** [comment]
@jchv The only file scanning there is now all in `FileScanner.cpp`. I am not sure whether this has now been fixed or not but in April we will be moving to C++17 where we will then use the `std::filesystem` API instead.

**Comment by AaronVanGeffen** [comment]
@IntelOrca I don't think `platform_enumerate_files_*`, `platform_enumerate_directories_*` use the FileScanner class yet.

**Comment by IntelOrca** [comment]
@AaronVanGeffen  https://github.com/OpenRCT2/OpenRCT2/pull/6969

**Comment by jchv** [comment]
Oh, that's great. So what do you think, wait for std::filesystem? I dunno if all of the code is elligable to use FileScanner, I am pretty sure I saw a lot of the old code in .c files.

**Comment by IntelOrca** [comment]
@jchv We are almost done with moving to C++, only a few .c files left. I'd say wait for `std::filesystem` since no user has reported this bug.

**Comment by tricktux** [comment]
- What I think needs to be done is to replace all the `platform_*` code with `std::filesystem`
	- the `platform_*` related with `std::filesystem`, like `platform_file_exists`
- Things to keep in mind:
	- Each platform looks for things in different places
- Approach:
	- [ ] create a FileSystem.{cpp.h}
	- [ ] implement `FileExists, GetUserDirectory, etc`
	- [ ] repeat the same thing `FileScanner.cpp`

- Related to #8032

**Comment by IntelOrca** [comment]
@tricktux The problem with `std::filesystem` has always been macOS' lack of it. We have however used boost to deal with this for OpenLoco.

https://github.com/OpenLoco/OpenLoco/blob/master/src/openloco/platform/platform.h

**Comment by tricktux** [comment]
@IntelOrca does that mean I should use `boost::filesystem`?

**Comment by duncanspumpkin** [comment]
It means we need to make a wrapper for boost:: filesystem and also make it so that mac builds include boost

**Comment by AaronVanGeffen** [comment]
Closing this, as the file referenced in the OP no longer matches the description.
